### ISSUE TITLE
[question] How to set border color and background color?

### ISSUE THREAD

**Issue by Fabrizio-Caruso** [z88dk/z88dk#763]
How can I set in the most portable way the background and the border colors?

Remark:
CC65 and WinCMOC provide:
(void) bordercolor(c)
(void) bgcolor (c);
as part of conio.h

**Comment by suborb** [comment]
I was planning on adding cborder(c) at some point, just a bit of a backlog at the moment!

**Comment by Fabrizio-Caruso** [comment]
@suborb for for the background?

If possible please use the same API as other 8-bit toolkits.

**Comment by zx70** [comment]
Very quick and dirty: this could be added to conio.h


/* Adding border colour where possible */
#ifdef __CPC__
#include <cpc.h>
#define bordercolor(a) cpc_SetBorder(a)
#endif

#ifdef __SPECTRUM__
#include <spectrum.h>
#define bordercolor(a) zx_border(a)
#endif

#ifdef __ZX81__
#include <zx81.h>
#define bordercolor(a) zx_border(a)
#endif

#ifdef __SMS__
#include <sms.h>
#define sms_border(a) sms_border_fastcall(a)
#endif


A better idea could be to write from scratch the new border functions but to refer to f_ansi_attr to decode the colors in a common way.

**Comment by Fabrizio-Caruso** [comment]
@zx70 and for the background color?

**Comment by zx70** [comment]
you may have noticed that there is already a problem with this trick for the border: color would change following the native colour rather than the conio ones.
with the background colour we will have another thing to consider: the new colour on some target would appear only after a cls, others would apply it immediately, in any case a similar trick is possible (with the same limit on the color encoding).
The 'better' idea I mentioned is a hint to save time if a new generic function will be written.

**Comment by suborb** [comment]
@Fabrizio-Caruso sorry, I thought cborder() was the standard one. Yes it'll make it standard.

@zx70 I'm keen not to drive everything down to lowest denominator so I'm minded to leave console and border as native colours and provide a (global) toggle to choose whether to map to a standard range. This way we provide both an enriched and a standard experience and the decision is in the hands of the developer which one they do.

**Comment by suborb** [comment]
@Fabrizio-Caruso a question about bgcolour() - it looks like cc65 uses it to set the screen to that colour immediately, whereas textbackground() (from conio.sf.net) sets it for characters moving forward.

So, how are you going to use it? If it's a case of setting the colour and then calling CLS then I'd implement textbackground() which should be pretty easy to do. 

If you want to set all the background to that colour then that's a fair bit more work and would be pretty slow on the VDP based machines since we'd have to read 8x bytes, mask, or, write back 8x bytes -  I don't think changing the palette in this case would make sense since we'd end up changing the foreground colour as well.

**Comment by Fabrizio-Caruso** [comment]
@suborb In my current use-case I need to set the background color before a clrscr, so textbackground() should probably do if clrscr fills the screen with space characters.
I guess (but I may be wrong) that textbackground() sets whatever background color each character should use.

**Comment by suborb** [comment]
That was the correct answer 😄 

And yes, your guess seems to be right, it'll be a shortcut for the VT52 sequence [ESC] b [value]

**Comment by suborb** [comment]
I've got a bit distracted by the Sord M5 but this is in progress. My current branch has the border colour changing on the TMS9918 targets, Spectrum and CPC. There's also a compile time option for disabling the colour mapping.

I need to continue to make sure the VT100 and VT52 terminals share the attribute variables so that I can remove the vtrendition() shortcut that's present in conio.h before merging.

**Comment by zx70** [comment]
If this makes you more comfortable, then wipe it immediately, I'm sure everyone will survive :)
We could open yet another issue listing the delta between ANSI VT only targets and the fully featured ones.

**Comment by suborb** [comment]
So now we have:

void textcolor(int c) __z88dk_fastcall;
void textbackground(int c) __z88dk_fastcall;

These should work on both the VT100 (ansi) and the VT52 (generic) consoles.

I've added:

void bordercolor(int c) __z88dk_fastcall;

This is available for all machines (it's a stub), but has actually been defined for ZX Spectrum, CPC and the TMS9918 platforms.
